sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+        headers: request.headers.toJSON()
+      };
+      if (proxySettings.username && proxySettings.password) {
+        proxyAgentOptions.auth = `${proxySettings.username}:${proxySettings.password}`;
+      }
+      proxyAgent = new httpsProxyAgent(proxyAgentOptions);
+    }
+    return proxyAgent;
+  } else if (request.keepAlive) {
+    if (!keepAliveAgent) {
+      keepAliveAgent = new https.Agent({
+        keepAlive: true
+      });
+    }
+
+    return keepAliveAgent;
+  } else {
+    return https.globalAgent;
+  }
+}
+
+function isReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+class ReportTransform extends Transform {
+  private loadedBytes = 0;
+  private progressCallback: (progress: TransferProgressEvent) => void;
+  _transform(chunk: string | Buffer, _encoding: string, callback: Function): void {
+    this.push(chunk);
+    this.loadedBytes += chunk.length;
+    this.progressCallback({ loadedBytes: this.loadedBytes });
+    callback();
+  }
+
+  constructor(progressCallback: (progress: TransferProgressEvent) => void) {
+    super();
+    this.progressCallback = progressCallback;
+  }
+}
+
+/**
+ * A HttpsClient implementation that uses Node's "https" module to send HTTPS requests.
+ */
+export class NodeHttpsClient implements HttpsClient {
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const abortController = new AbortController();
+    let abortListener: ((event: any) => void) | undefined;
+    if (request.abortSignal) {
+      if (request.abortSignal.aborted) {
+        throw new AbortError("The operation was aborted.");
+      }
+
+      abortListener = (event: Event) => {
+        if (event.type === "abort") {
+          abortController.abort();
+        }
+      };
+      request.abortSignal.addEventListener("abort", abortListener);
+    }
+
+    if (request.timeout > 0) {
+      setTimeout(() => {
+        abortController.abort();
+      }, request.timeout);
+    }
+
+    if (request.formData) {
+      prepareFormData(request.formData, request);
+    }
+
+    if (!request.skipDecompressResponse) {
+      request.headers.set("Accept-Encoding", "gzip,deflate");
+    }
+
+    let body = request.body;
+
+    if (body && request.onUploadProgress) {
+      const onUploadProgress = request.onUploadProgress;
+      const uploadReportStream = new ReportTransform(onUploadProgress);
+      if (isReadableStream(body)) {
+        body.pipe(uploadReportStream);
+      } else {
+        uploadReportStream.end(body);
+      }
+
+      body = uploadReportStream;
+    }
+
+    if (body) {
+      if (isReadableStream(body)) {
+        request.headers.set("Transfer-Encoding", "chunked");
+      } else if (typeof body === "string") {
+        request.headers.set("Content-Length", body.length);
```

Comment:
Is content-length something that makes sense as a policy as well, that happens after serialization, just to cut down on what's happening in the http client (or do we want the http client to do more?)

I do think this needs to calculate the content-length of more than just string though. Some services (I'm thinking of storage-blob) may need to know the content-length of a request up-front. If the user passed in anything other than a stream (e.g. Blob, Buffer), I'd expect we can set the Content-Length.

If the user passed in a file-based stream we could use the fs API to get the length of the file as well if it wasn't provided!

Reply:
I cleaned this up a little bit. I didn't do anything special for file streams, though. Is there some magic way to know the length of a file stream? I thought you had to stat the file to get that kind of info?

Reply:
That would be the magic way of knowing :) (I'm not suggesting we do that now, just pointing out that if this were in a policy, we could conditionally enable this if we wanted to without the http client needing to know anything about how the content-length was set.

Reply:
that would be great!